beacon_node/operation_pool/src/lib.rs

```diff
@@ -295,27 +305,26 @@ impl<T: EthSpec> OperationPool<T> {
         )
     }
 
-    /// Prune if validator has already exited at the last finalized state.
-    pub fn prune_voluntary_exits(&self, finalized_state: &BeaconState<T>, spec: &ChainSpec) {
+    /// Prune if validator has already exited at or before the finalized checkpoint of the head.
+    pub fn prune_voluntary_exits(&self, head_state: &BeaconState<T>) {
         prune_validator_hash_map(
             &mut self.voluntary_exits.write(),
-            |validator| validator.exit_epoch != spec.far_future_epoch,
-            finalized_state,
+            // This condition is slightly too loose, since there will be some finalized exits that
+            // are missed here.
+            //
+            // We choose simplicity over the gain of pruning more exits since they are small and
+            // should not be seen frequently.
+            |validator| validator.exit_epoch <= head_state.finalized_checkpoint.epoch,
+            head_state,
         );
     }
 
-    /// Prune all types of transactions given the latest finalized state and head fork.
-    pub fn prune_all(
-        &self,
-        finalized_state: &BeaconState<T>,
-        current_epoch: Epoch,
-        head_fork: Fork,
-        spec: &ChainSpec,
-    ) {
+    /// Prune all types of transactions given the latest head state and head fork.
```

Comment:
```suggestion
    /// Prune all types of transactions given the latest head state and current epoch.
```

Reply:
We could also derive the current epoch from the head state to avoid passing parameters around.

Reply:
I think we want to keep `current_epoch` since it's the wall-clock epoch.

Reply:
Yes!!